### PR TITLE
Reduce memory overhead for webpack configs with many entries

### LINKED ISSUE #143: Process aborts with 'out of memory' when using 2.0.0
- Operating System: OSX 10.14.6
- Node Version: 10.16.0
- NPM Version: 6.9.0
- webpack Version: 4.39.3
- terser-webpack-plugin Version: 2.0.0

### Expected Behavior

Process does not abort

### Actual Behavior

```
$ NODE_ENV=production ./node_modules/.bin/webpack --loglevel notice

<--- Last few GCs --->

[84294:0x102843000]    55749 ms: Mark-sweep 1312.4 (1444.3) -> 1305.0 (1446.3) MB, 622.0 / 0.0 ms  (average mu = 0.099, current mu = 0.040) allocation failure scavenge might not succeed
[84294:0x102843000]    56388 ms: Mark-sweep 1315.7 (1446.3) -> 1308.6 (1448.8) MB, 613.1 / 0.0 ms  (average mu = 0.070, current mu = 0.040) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x3aa9ebf5be3d]
Security context: 0x37c8eea1e6e9 <JSObject>
    1: /* anonymous */(aka /* anonymous */) [0x37c895d8d969] [/Users/nick/Projects/origin/node_modules/terser-webpack-plugin/node_modules/webpack-sources/lib/applySourceMap.js:~58] [pc=0x3aa9ed4fe2c4](this=0x37c83a5026f1 <undefined>,chunk=0x37c803a7d979 <String[14]: createElement(>,middleMapping=0x37c8171046d9 <Object map = 0x37c835e45c79>)
    2: SourceNode...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10003cf99 node::Abort() [/usr/local/bin/node]
 2: 0x10003d1a3 node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 3: 0x1001b7835 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 4: 0x100585682 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 5: 0x100588155 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/usr/local/bin/node]
 6: 0x100583fff v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 7: 0x1005821d4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
 8: 0x10058ea6c v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
 9: 0x10058eaef v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
10: 0x10055e434 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/usr/local/bin/node]
11: 0x1007e6714 v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
12: 0x3aa9ebf5be3d 
13: 0x3aa9ed4fe2c4 
14: 0x3aa9ed4b7d28 
Abort trap: 6
```

### Code

https://github.com/OriginProtocol/origin/blob/master/dapps/marketplace/webpack.config.js

### How Do We Reproduce?

```
git clone https://github.com/OriginProtocol/origin.git
cd origin
# Update dapps/marketplace/package.json to use v2.0.0 of terser plugin
yarn
cd dapps/marketplace
yarn build
```

### PR DESCRIPTION
**Pull request by cjlarose**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

This change dramatically reduces the total size of objects on the heap at any one time for webpack configurations with many entries. Because `tasks` used to be computed eagerly and `completedTasks` was collected as a large Array, space complexity used to be linear with respect to the number of entries. (More precisely, the max heap size was proportional to the combined size in bytes of all pre-minified sources).

Now, we defer the generation of a `task` (and thus the memory allocation required for computing `asset.source()` of an entry) until a worker is made available. Similarly, the computation of `serialize(task)`, another large String, is deferred until a worker is available. Finally, when a
`task` is completed, the `completedTask` is consumed immediately, releasing the reference to the original `asset.source()`, and making it a candidate for garbage collection.

The effect is that space complexity is now roughly linear with respect to the number of parallel workers. The number of pre-minified sources that are considered "live" to the garbage collector will never exceed the number of parallel workers.

I believe this fixes #143 for most people experiencing the problem.

I created a repository that reliably reproduces the issue: https://github.com/cjlarose/terser-webpack-plugin-out-of-memory. Using the changes introduced in this PR, that project builds successfully using the default Node old space size (1400MB).

### Breaking Changes

None

### CODE REVIEW

**Comment by tannera** [comment]
Looking forward to this!

File: src/index.js [hunk 1]
```diff
@@ -184,13 +184,7 @@ class TerserPlugin {
...
 
   static hasAsset(commentFilename, assets) {
-    const assetFilenames = Object.keys(assets).map((assetFilename) =>
-      TerserPlugin.removeQueryString(assetFilename)
-    );
-
-    return assetFilenames.includes(
-      TerserPlugin.removeQueryString(commentFilename)
-    );
+    return assets.has(TerserPlugin.removeQueryString(commentFilename));
```
Comment (evilebottnawi) [inline src/index.js]: Why it was changed, refactor? Please avoid not relevant changes

File: src/index.js [hunk 2]
```diff
@@ -249,145 +243,151 @@ class TerserPlugin {
         []
       );
       const files = [].concat(additionalChunkAssets).concat(chunksFiles);
-      const tasks = [];
 
-      files.forEach((file) => {
-        if (!matchObject(file)) {
-          return;
-        }
+      function* tasks() {
```
Comment (evilebottnawi) [inline src/index.js]: Move this to class method, for better readable

File: src/TaskRunner.js [hunk 3]
```diff
@@ -40,8 +40,26 @@ export default class TaskRunner {
...
+          onCompletedTask(task, completedTask);
+          delete handlers[idx];
+          resolve();
+        };
+        while (outputIndex in handlers) {
+          handlers[outputIndex]();
+          outputIndex += 1;
+        }
+      });
+    };
```
Comment (evilebottnawi) [inline src/TaskRunner.js]: Looks like it is kill idea of parallels, because you run next task after other task

File: src/index.js [hunk 4]
```diff
@@ -249,145 +243,151 @@ class TerserPlugin {
...
+        const existingAssets = new Set(
```
Comment (evilebottnawi) [inline src/index.js]: Not sure it is good idea, one each task we will create new Set, in webpack@5 it is already `Set` so better uses `compilation.assets`

**Comment by cjlarose** [comment]
I collected benchmarks using the repo I created that demonstrates the original `JavaScript heap out of memory` bug (https://github.com/cjlarose/terser-webpack-plugin-out-of-memory/tree/cf9fc24cee6b2085a472de3b36a37be7210238e7)

I executed builds using my Mid-2017 MacBook Pro (Intel Core i7-7820HQ Quad-Core 2.9GHz), with cache off and cache on, with 4 terser workers. I'm using macOS 10.14.6 and Node 12.14.1.

"Fails" in this table represents somewhere where `terser-webpack-plugin` version 2.3.2 failed with an out-of-memory error.


| # entries | cache | max-old-space-size | 2.3.2 (s) | this PR  (s) |
|-----------|-------|--------------------|-----------|--------------|
| 10        | cold  | 1400MB             | 4.17      | 4.76         |
| 10        | warm  | 1400MB             | 1.05      | 1.19         |
| 100       | cold  | 1400MB             | 23.15     | 23.76        |
| 100       | warm  | 1400MB             | 1.77      | 1.98         |
| 500       | cold  | 1400MB             | Fails     | 104.42       |
| 500       | warm  | 1400MB             | 5.10      | 6.38         |
| 500       | cold  | 4096MB             | 109.32    | 103.79       |
| 500       | warm  | 4096MB             | 5.13      | 6.02         |
| 1000      | cold  | 1400MB             | Fails     | 214.10       |
| 1000      | warm  | 1400MB             | 10.50     | 10.42        |
| 1000      | cold  | 4096MB             | Fails     | 242.01       |
| 1000      | warm  | 4096MB             | 10.68     | 11.53        |

In conclusion, there are some configurations for which this change introduces a minor performance regression. This is mostly expected since we're basically forcing the "main" process to wait until work is completed before generating the next task in order to conserve memory. The takeaway, though, is that this change makes it possible to build webpack configurations with large numbers of entries reliably.

File: src/index.js (hunk 2 above)
Comment (cjlarose) [reply src/index.js]: Good idea! I moved this to a separate method as part of a separate commit fa36e0f

File: src/index.js (hunk 1 above)
Comment (cjlarose) [reply src/index.js]: I'm sorry that I didn't make clear why this change is necessary. I extracted this change into a separate commit 35f310c0a22da98bc459a7d8729ff1cb2879c2bc and explain why it's necessary there. Let me know if it's still not clear.

File: src/index.js (hunk 4 above)
Comment (cjlarose) [reply src/index.js]: I don't think this is right. Only one `Set` is generated for a `compilation` (not for every task). In fact, before this change was made, a new `Array` was generated for every task. I explain this more in the commit 35f310c0a22da98bc459a7d8729ff1cb2879c2bc

`compilation.assets` is not a `Set` in webpack@5. You might be thinking of `compilation.modules` or `compilation.chunks`.

File: src/index.js [hunk 5]
```diff
@@ -184,19 +184,175 @@ class TerserPlugin {
...
+          const data = { filename, basename, query };
+
+          commentsFilename = compilation.getPath(commentsFilename, data);
+        }
+
+        if (
+          commentsFilename &&
+          TerserPlugin.hasAsset(commentsFilename, existingAssets)
+        ) {
+          // Todo make error and stop uglifing in next major release
```
Comment (jamesgeorge007) [inline src/index.js]:
```suggestion
          // Todo make error and stop uglifying in next major release
```

**Comment by avshalomt2** [comment]
@cjlarose Do you believe this fix solves the same error when using `terser-webpack-plugin-legacy` (based on branch `webpack-3`)?
I just looked at the code in the `webpack-3` branch and it seems possible to backport this fix to there, and i'm thinking about doing so.

@evilebottnawi If I create such a PR, could it be merge to the currently stale `webpack-3` branch?

**Comment by evilebottnawi** [comment]
> If I create such a PR, could it be merge to the currently stale webpack-3 branch?

No, please migrate on latest version

**Comment by evilebottnawi** [comment]
WIP on tomorrow :+1:

**Comment by evilebottnawi** [comment]
@cjlarose Can you fix conflicts?

File: src/index.js (hunk 1 above)
Comment (evilebottnawi) [reply src/index.js]: It is expected, two difference tasks should not generate same files

File: src/index.js [hunk 6]
```diff
@@ -184,19 +184,175 @@ class TerserPlugin {
...
-    );
+    return assets.has(TerserPlugin.removeQueryString(commentFilename));
   }
 
   static isWebpack4() {
     return webpackVersion[0] === '4';
   }
 
+  *generateTasks(compiler, compilation, chunks, processedAssets) {
+    const existingAssets = new Set(
```
Comment (evilebottnawi) [inline src/index.js]: We really do not need Set here, assets can very large, so we should avoid create Set, also avoiding Set should decrease CPU and memory usage

File: src/TaskRunner.js [hunk 7]
```diff
@@ -61,13 +79,22 @@ export default class TaskRunner {
...
-        return enqueue();
-      })
-    );
+        // eslint-disable-next-line no-await-in-loop
+        await offerResult(inputIndex, task, await promise);
+      }
+    };
+
+    const workerPromises = [];
+    for (let i = 0; i < Math.max(1, this.numberWorkers); i++) {
```
Comment (evilebottnawi) [inline src/TaskRunner.js]: `parallel: false` is broken in that case

File: src/TaskRunner.js [hunk 8]
```diff
@@ -61,13 +79,22 @@ export default class TaskRunner {
...
+        const promise = this.cache.isEnabled()
+          ? this.cache.get(task).then((data) => data, enqueue)
+          : enqueue();
+
+        inputIndex += 1;
 
-        return enqueue();
-      })
-    );
+        // eslint-disable-next-line no-await-in-loop
```
Comment (evilebottnawi) [inline src/TaskRunner.js]: Not good idea, need refactor

**Comment by evilebottnawi** [comment]
I rewrited your idea like JavaScript Beginner Book :smile: Big thanks for a inspiration! It was great idea. Perhaps you still have some ideas for optimization, so I’ll leave this PR open for a while, feel free to leave feedback

**Comment by cjlarose** [comment]
I'll close it for now. I can introduce separate optimizations in new PRs if necessary. Thanks.

**Comment by evilebottnawi** [comment]
@cjlarose Great! Will be wait it :+1:

**Comment by cjlarose** [comment]
This work was re-implemented in #211
